### PR TITLE
Trampoline typeclasses

### PR DESCRIPTION
**Pull request by ferranpujolcamins**
## Breaking changes
The result of `map` and `flatMap` now needs to be fixed with `^`. 

## Related issues

#625

### CODE REVIEW

File: Sources/Bow/Data/Trampoline.swift [hunk 1]
```diff
@@ -40,78 +54,136 @@ public class Trampoline<A> {
...
+        }
     }
 }
 
-private final class Defer<A>: Trampoline<A> {
-    let deferred: () -> Trampoline<A>
+private final class FlatMapStep<A>: CokleisliK<CoyonedaFPartial<ForTrampoline, Trampoline<A>>, Either<() -> Trampoline<A>, A>> {
+    override init() {}
+
+    override func invoke<X>(_ fa: Kind<CoyonedaFPartial<ForTrampoline, Trampoline<A>>, X>) -> Either<() -> Trampoline<A>, A> {
```
Comment (truizlop) [inline Sources/Bow/Data/Trampoline.swift:118]:
```suggestion
    override func invoke<X>(_ fa: CoyonedaFOf<ForTrampoline, Trampoline<A>, X>) -> Either<() -> Trampoline<A>, A> {
```

**Review by truizlop — APPROVED** [review]
Everything looks good to me. I see you have moved Coyoneda to the main module; I am wondering if we should move Yoneda too, to maintain consistency, and keep only Free and Cofree in BowFree. What do you think?

**Comment by ferranpujolcamins** [comment]
> Everything looks good to me. I see you have moved Coyoneda to the main module; I am wondering if we should move Yoneda too, to maintain consistency, and keep only Free and Cofree in BowFree. What do you think?

Sounds good to me. Let me do that in another PR.

**Comment by ferranpujolcamins** [comment]
@truizlop can this be merged?

**Pull request merged**